src/cascadia/TerminalApp/AppLogic.cpp

```diff
@@ -1447,6 +1470,40 @@ namespace winrt::TerminalApp::implementation
         return _settings.GlobalSettings().ActionMap().GlobalHotkeys();
     }
 
+    bool AppLogic::ShouldUsePersistedLayout()
+    {
+        return _root != nullptr ? _root->ShouldUsePersistedLayout(_settings) : false;
+    }
+
+    void AppLogic::SaveWindowLayoutJsons(Windows::Foundation::Collections::IVector<hstring> layouts)

```

Comment:
Passing it by reference would be better.

From an architectural standpoint I dislike that we pass JSON strings between classes and I'd prefer if we'd use `WindowLayout` or something similar. This is my pet peeve with this PR.

Reply:
I did strings because they need to go through the peasant/monarch and the remoting lib does not reference `Terminal::Settings::Model`.

Reply:
> Passing it by reference would be better.

Because I have this function defined in the IDL, I think I have to keep it like this?

Reply:
> I did strings because they need to go through the peasant/monarch and the remoting lib does not reference Terminal::Settings::Model.

That makes sense. It's quite unfortunate however...

> Because I have this function defined in the IDL, I think I have to keep it like this?

Nope! The cppwinrt compiler generates the actual classes (e.g. `winrt::AppLogic`). You just write the `winrt::implementation::AppLogic` class. For each function defined in the .idl, the cppwinrt compiler generates a wrapper function like this:
```cpp
void __stdcall SaveWindowLayoutJsons(void** value) noexcept final try
{
    typename D::abi_guard guard(this->shim());
    this->shim().SaveWindowLayoutJsons(*reinterpret_cast<winrt::Windows::Foundation::Collections::IVector<winrt::hstring> const*>(&value));
}
```

As you can see you could theoretically even have a "incompatible" function definition, as long as it can be implicitly cast. Taking arguments by reference is always possible with this.

Reply:
Hmm, maybe I just did something wrong because I got a compiler error when trying to add the `&` version and in my hurry to get to everything else didn't really investigate what was wrong exactly. I can try again.

Reply:
had to be a `const &` not just a `&` :shrug: